extensions/amp-brightcove/0.1/amp-brightcove.js

```diff
@@ -39,21 +82,128 @@ class AmpBrightcove extends AMP.BaseElement {
     return isLayoutSizeDefined(layout);
   }
 
+  /** @override */
+  viewportCallback(visible) {
+    this.element.dispatchCustomEvent(VideoEvents.VISIBILITY, {
+      visible,
+    });
+  }
+
   /** @override */
   buildCallback() {
     this.iframe_ = null;
+
+    this.playerReadyPromise_ = new Promise(resolve => {
+      this.playerReadyResolver_ = resolve;
+    });
+
+    // Warn if the player does not have video interface support
+    this.readyTimeout_ = window.setTimeout(() => {
+      dev().warn(TAG,
+          `Did not receive ready callback from player ${this.playerId_}.` +
+        ' Ensure it has the videojs-amp-support plugin configured.');
+    }, 3000);
+
+    this.playerReadyResolver_(this.iframe_);
   }
 
   /** @override */
   layoutCallback() {
     const iframe = this.element.ownerDocument.createElement('iframe');
+
     iframe.setAttribute('frameborder', '0');
     iframe.setAttribute('allowfullscreen', 'true');
     iframe.src = this.getIframeSrc_();
+
     this.applyFillContent(iframe);
     this.element.appendChild(iframe);
     this.iframe_ = iframe;
-    return this.loadPromise(iframe);
+
+    this.unlistenMessage_ = listen(
+        this.win,
+        'message',
+        this.handlePlayerMessages_.bind(this)
+    );
+
+    return this.loadPromise(iframe)
+        .then(() => this.playerReadyPromise_);
+  }
+
+  /**
+   * Sends a command to the player through postMessage.
+   * @param {string} command
+   * @param {*=} arg
+   * @private
+   * */
+  sendCommand_(command, arg) {
+    this.iframe_.contentWindow. /*OK*/ postMessage(JSON.stringify(dict({
+      'command': command,
+      'args': arg,
+    })), 'https://players.brightcove.net');
+  }
+
+  /** @private */
+
+  handlePlayerMessages_(event) {
+    if (event.origin != 'https://players.brightcove.net' ||
```

Comment:
nit: Assign `this.element` to a variable in every place where you use it more than once. Makes the file slightly smaller.